src/test/java/com/example/bookshop/service/impl/BookServiceImplTest.java

```diff
@@ -0,0 +1,259 @@
+package com.example.bookshop.service.impl;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+import com.example.bookshop.dto.book.BookDto;
+import com.example.bookshop.dto.book.BookToDtoWithoutCategoryIds;
+import com.example.bookshop.dto.book.CreateBookRequestDto;
+import com.example.bookshop.exception.EntityNotFoundException;
+import com.example.bookshop.mapper.BookMapper;
+import com.example.bookshop.mapper.impl.BookMapperImpl;
+import com.example.bookshop.model.Book;
+import com.example.bookshop.model.Category;
+import com.example.bookshop.repository.book.BookRepository;
+import com.example.bookshop.repository.category.CategoryRepository;
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mockito;
+import org.mockito.Spy;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.PageImpl;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+
+@ExtendWith(MockitoExtension.class)
+class BookServiceImplTest {
+    @InjectMocks
+    private BookServiceImpl bookService;
+    @Spy
+    private BookRepository bookRepository;

```

Comment:
```suggestion
    @Mock
    private BookRepository bookRepository;
```
Use Spy only with Mappers